### PR TITLE
YD-297 Fixed issues around partially established buddy relationships

### PR DESCRIPTION
**Pull request by Bert-R**

### CODE REVIEW

File: core/src/main/java/nu/yona/server/analysis/service/ActivityService.java [hunk 1]
```diff
@@ -175,7 +175,18 @@ private long getTotalPageableItems(UserAnonymizedDTO userAnonymized, ChronoUnit
 	@Transactional
 	public Page<DayActivityOverviewDTO<DayActivityDTO>> getBuddyDayActivityOverviews(UUID buddyID, Pageable pageable)
 	{
-		return getDayActivityOverviews(buddyService.getBuddy(buddyID).getUserAnonymizedID(), pageable);
+		return getDayActivityOverviews(getUserAnonymizedID(buddyID), pageable);
+	}
+
+	private UUID getUserAnonymizedID(UUID buddyID)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/analysis/service/ActivityService.java]: Please rename to `getUserAnonymizedIDOfBuddy`

File: core/src/main/java/nu/yona/server/analysis/service/ActivityService.java [hunk 2]
```diff
@@ -175,7 +175,18 @@ private long getTotalPageableItems(UserAnonymizedDTO userAnonymized, ChronoUnit
...
+	{
+		return getUserAnonymizedID(buddyService.getBuddy(buddyID));
+	}
+
+	private UUID getUserAnonymizedID(BuddyDTO buddy)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/analysis/service/ActivityService.java]: Here too maybe rename to `getUserAnonymizedIDOfBuddy`?

File: appservice/src/main/java/nu/yona/server/subscriptions/rest/BuddyController.java [hunk 3]
```diff
@@ -291,8 +291,11 @@ public BuddyResource toResource(BuddyDTO buddy)
 			ControllerLinkBuilder selfLinkBuilder = getSelfLinkBuilder(buddy.getID());
 			addSelfLink(selfLinkBuilder, buddyResource);
 			addEditLink(selfLinkBuilder, buddyResource);
-			addDayActivityOverviewsLink(buddyResource);
-			addWeekActivityOverviewsLink(buddyResource);
+			if (buddy.getSendingStatus() == Status.ACCEPTED)
```
Comment (aristotelos) [inline appservice/src/main/java/nu/yona/server/subscriptions/rest/BuddyController.java]: Could you add a javadoc comment to explain this is the buddy who sends (and not the user having the buddy)? In the Swagger spec too (two places).

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java [hunk 4]
```diff
@@ -302,9 +373,12 @@ void sendBuddyConnectResponseMessage(UUID senderUserID, UUID senderUserAnonymize
 
 	private void disconnectBuddy(UserAnonymizedDTO userAnonymized, UUID userAnonymizedID)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Consider to rename to `disconnectBuddyIfPresent`. By the way, how can it be that the `buddyAnonymized` is not present while the sending and receiving status is `ACCEPTED` (as tested in the caller of this method)?

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java [hunk 5]
```diff
@@ -170,6 +186,11 @@ public void removeBuddy(UUID idOfRequestingUser, UUID buddyID, Optional<String>
 		User user = userService.getValidatedUserbyID(idOfRequestingUser);
 		Buddy buddy = getEntityByID(buddyID);
 
+		if (buddy.getSendingStatus() == Status.REQUESTED || buddy.getReceivingStatus() == Status.REQUESTED)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Could you add a comment here why this is necessary?

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java [hunk 6]
```diff
@@ -221,18 +277,29 @@ void removeBuddyInfoForRemovedUser(User user, Buddy buddy)
...
-			sendBuddyConnectResponseMessage(null, buddy.getUserAnonymizedID(), buddy.getNickname(), user.getUserAnonymizedID(),
+			// Send message to "self", as if the requested user declined the buddy request
+			UUID buddyUserAnonymizedID = buddy.getUserAnonymizedID().orElse(null); //
+			sendBuddyConnectResponseMessage(null, buddyUserAnonymizedID, buddy.getNickname(), user.getUserAnonymizedID(),
 					buddy.getID(), Status.REJECTED, getDropBuddyMessage(DropBuddyReason.USER_ACCOUNT_DELETED, Optional.empty()));
 		}
 		removeBuddy(user, buddy);
 	}
 
+	private UUID getUserAnonymizedIDForBuddy(Buddy buddy, String state)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: I always see the same `state` string in the code. Should it not discriminate between the different occasions?

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy [hunk 7]
```diff
@@ -533,4 +536,189 @@ class BasicBuddyTest extends AbstractAppServiceIntegrationTest
...
+		response.status == 200
+		appService.getBuddies(bob).size() == 1 // Buddy not yet removed for Bob (didn't process the disconnect)
+		appService.getBuddies(richard).size() == 0 // Buddy removed for Richard
+
+		appService.getMessages(bob).responseData.page.totalElements == 1 // Only the disconnect message
+
+		def disconnectMessage = appService.getMessages(bob).responseData._embedded."yona:messages".findAll{ it."@type" == "BuddyDisconnectMessage"}[0]
+		def processURL = disconnectMessage._links."yona:process".href
+
+		when:
```
Comment (aristotelos) [inline appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: Second `when`

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy [hunk 8]
```diff
@@ -533,4 +536,189 @@ class BasicBuddyTest extends AbstractAppServiceIntegrationTest
...
+		appService.postMessageActionWithPassword(acceptURL, ["message" : "Yes, great idea!"], bob.password)
+
+		when:
+		def response = appService.removeBuddy(bob, appService.getBuddies(bob)[0], "Sorry, I regret accepting you")
+
+		then:
+		response.status == 200
+		appService.getBuddies(bob).size() == 0 // Buddy removed for Bob
+		def buddiesRichard = appService.getBuddies(richard)
+		buddiesRichard.size() == 1 // Buddy removed for Richard
```
Comment (aristotelos) [inline appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: `// Buddy not yet removed for Richard`

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy [hunk 9]
```diff
@@ -533,4 +536,189 @@ class BasicBuddyTest extends AbstractAppServiceIntegrationTest
...
+
+		def responseGetMessagesUser = appService.getMessages(user)
+		assert responseGetMessagesUser.status == 200
+		def goalConflictMessagesUser = responseGetMessagesUser.responseData._embedded."yona:messages".findAll{ it."@type" == "GoalConflictMessage"}
+		assert goalConflictMessagesUser.size() == 1
+		def responseDeleteMessageUser = appService.deleteResourceWithPassword(goalConflictMessagesUser[0]._links.edit.href, user.password)
+		responseDeleteMessageUser.status == 200
+	}
+
+	private void assertGoalConflictIsNotReportedToBuddy(User user, User buddy)
```
Comment (aristotelos) [inline appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: Method is not yet used? Can be added after `disconnectBuddy` calls.

File: appservice/src/intTest/groovy/nu/yona/server/ActivityTest.groovy [hunk 10]
```diff
@@ -460,6 +460,86 @@ class ActivityTest extends AbstractAppServiceIntegrationTest
...
+		richard.buddies[0].weeklyActivityReportsUrl == null
+		responseDayOverviewsWithBuddies.status == 200
+		responseDayOverviewsWithBuddies.responseData._embedded."yona:dayActivityOverviews"[0].dayActivities.find{ it._links."yona:activityCategory"?.href == GAMBLING_ACT_CAT_URL}.dayActivitiesForUsers.size() == 1
+
+		cleanup:
+		appService.deleteUser(richard)
+		appService.deleteUser(bob)
+	}
+
+	def 'Richard retrieves buddy activity info before he processed Bob\'s acceptance'()
```
Comment (aristotelos) [inline appservice/src/intTest/groovy/nu/yona/server/ActivityTest.groovy]: Maybe it is also good to add a test for the scenario that Bob removed the buddy entry, or that Richard removed the buddy entry before the process?

**Comment by aristotelos** [comment]
Added a few comments/questions.

File: core/src/main/java/nu/yona/server/analysis/service/ActivityService.java (hunk 1 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/analysis/service/ActivityService.java]: Done: ``getBuddyUserAnonymizedID``

File: core/src/main/java/nu/yona/server/analysis/service/ActivityService.java (hunk 2 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/analysis/service/ActivityService.java]: Done: ``getBuddyUserAnonymizedID``

File: appservice/src/main/java/nu/yona/server/subscriptions/rest/BuddyController.java (hunk 3 above)
Comment (Bert-R) [inline appservice/src/main/java/nu/yona/server/subscriptions/rest/BuddyController.java]: Done

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java (hunk 4 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Done: ``disconnectBuddyIfConnected``, and also renamed ``userAnonymized`` into ``buddyUserAnonymized``.

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java (hunk 5 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Done.

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java (hunk 6 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java]: Removed the parameter and extended the fixed string.

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy (hunk 7 above)
Comment (Bert-R) [inline appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: Removed.

File: core/src/main/java/nu/yona/server/subscriptions/entities/BuddyAnonymized.java [hunk 11]
```diff
@@ -30,7 +30,16 @@ public static BuddyAnonymizedRepository getRepository()
...
 	private UUID userAnonymizedID;
 
+	/*
+	 * When the sendingStatus is accepted, the buddy user ACCEPTED to send goal conflicts and activity information to the other
+	 * side.
+	 */
 	private Status sendingStatus = Status.NOT_REQUESTED;
+
+	/*
+	 * When the receivingStatus is accepted, the user ACCEPTED to receive goal conflicts and activity information from the other
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/entities/BuddyAnonymized.java]: "the buddy user"

**Pull request merged**